### PR TITLE
UCP/WIREUP: Generalize lane_usage as "lane_type"

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -59,11 +59,13 @@ static ucs_stats_class_t ucp_ep_stats_class = {
 void ucp_ep_config_key_reset(ucp_ep_config_key_t *key)
 {
     ucp_lane_index_t i;
+
     memset(key, 0, sizeof(*key));
     key->num_lanes        = 0;
     for (i = 0; i < UCP_MAX_LANES; ++i) {
         key->lanes[i].rsc_index    = UCP_NULL_RESOURCE;
         key->lanes[i].proxy_lane   = UCP_NULL_LANE;
+        key->lanes[i].lane_types    = 0;
```
Comment (dmitrygx) [inline src/ucp/core/ucp_ep.c:68]: unaligned by `=`

File: src/ucp/core/ucp_ep.h [hunk 2]
```diff
@@ -107,12 +108,14 @@ typedef struct ucp_ep_config_key {
     ucp_lane_index_t       num_lanes;    /* Number of active lanes */
 
     struct {
-        ucp_rsc_index_t    rsc_index;    /* Resource index */
-        ucp_lane_index_t   proxy_lane;   /* UCP_NULL_LANE - no proxy
+        ucp_rsc_index_t      rsc_index;    /* Resource index */
+        ucp_lane_index_t     proxy_lane;   /* UCP_NULL_LANE - no proxy
                                             otherwise - in which lane the real
```
Comment (dmitrygx) [inline src/ucp/core/ucp_ep.h:113]: unaligned

File: src/ucp/proto/lane_type.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        .short_name = "am_bw"
+    },
+    [UCP_LANE_TYPE_RMA] = {
+        .short_name = "rma"
+    },
+    [UCP_LANE_TYPE_RMA_BW] = {
+        .short_name = "rma_bw"
+    },
+    [UCP_LANE_TYPE_RKEY_PTR] = {
+        .short_name = "rk_ptr"
```
Comment (dmitrygx) [inline src/ucp/proto/lane_type.c:30]: would it be better to use `"rkey_ptr"` instead?

File: src/ucp/proto/lane_type.c [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    [UCP_LANE_TYPE_AMO] = {
+        .short_name = "amo"
+    },
+    [UCP_LANE_TYPE_TAG] = {
+        .short_name = "tag"
+    },
+    [UCP_LANE_TYPE_CM] = {
+        .short_name = ""
```
Comment (dmitrygx) [inline src/ucp/proto/lane_type.c:39]: q: `"cm"`?
or why we have empty string here?

File: src/ucp/wireup/select.c [hunk 5]
```diff
@@ -543,25 +530,14 @@ ucp_wireup_add_lane_desc(const ucp_wireup_select_info_t *select_info,
     lane_desc->path_index   = select_info->path_index;
     lane_desc->proxy_lane   = proxy_lane;
     lane_desc->dst_md_index = dst_md_index;
-    lane_desc->usage        = usage;
-    lane_desc->am_bw_score  = 0.0;
-    lane_desc->rma_score    = 0.0;
-    lane_desc->rma_bw_score = 0.0;
-    lane_desc->amo_score    = 0.0;
+    lane_desc->lane_types    = UCS_BIT(lane_type);
```
Comment (dmitrygx) [inline src/ucp/wireup/select.c:533]: unaligned by `=`

File: src/ucp/wireup/select.c [hunk 6]
```diff
@@ -598,44 +574,46 @@ ucp_wireup_add_lane(const ucp_wireup_select_params_t *select_params,
...
-        \
-        score1 = (*lane1 == UCP_NULL_LANE) ? 0.0 : lanes[*lane1]._token##_score; \
-        score2 = (*lane2 == UCP_NULL_LANE) ? 0.0 : lanes[*lane2]._token##_score; \
-        /* sort from highest score to lowest */ \
-        (score1 < score2) ? 1 : ((score1 > score2) ? -1 : 0); \
-    })
+static int ucp_wireup_compare_score(const void *elem1, const void *elem2,
+                                    void *arg, ucp_lane_type_t lane_type)
+{
+    const ucp_lane_index_t *lane1 = elem1;
```
Comment (dmitrygx) [inline src/ucp/wireup/select.c:583]: align by `=`?

**Comment by yosefe** [comment]
@dmitrygx thanks for the comments! fixed

**Comment by brminich** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**